### ISSUE TITLE
rails-ujs not working with remote: true Ajax request 

### ISSUE THREAD

**Issue by kirykr** [rails/rails#33115]
### Steps to reproduce
In my `app/assets/javascripts/admin.js`
```
//= require jquery
//= require rails-ujs
//= require activestorage
```
in my view link `app/views/admin/categories/index.haml`
```
= link_to admin_category_path(category), remote: true, method: 'delete', data: { confirm: 'Are you sure you want to delete?', turbolinks: false, disable_with: 'Processing...'} do
  %i.fa.fa-trash
```
in my `app/controllers/admin/categories_controller.rb`
```ruby
 def destroy
      sleep 1
      @category = @category.destroy
      respond_to do |format|
        format.html { redirect_to admin_categories_url, notice: 'Category was successfully destroyed.' }
        format.json { head :no_content }
        format.js
      end
    end

    private
  
      def set_category
        @category = Category.friendly.find(params[:id])
      end
```
in my `app/views/admin/categories/destroy.js.erb`
```
$("category_<%= @category.id %>").remove();
```
(Guidelines for creating a bug report are [available
here](http://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#creating-a-bug-report))

### Expected behavior
The table row to be removed.

### Actual behavior
the data was removed from my database successfully
but the table row was not removed because of the error below
```
Uncaught SyntaxError: Unexpected token <
    at processResponse (rails-ujs.self-551fbd47b981dacbb84a270f9123074caf39eb72aaf6f478ab597c6f81435e4b.js?body=1:268)
    at rails-ujs.self-551fbd47b981dacbb84a270f9123074caf39eb72aaf6f478ab597c6f81435e4b.js?body=1:196
    at XMLHttpRequest.xhr.onreadystatechange (rails-ujs.self-551fbd47b981dacbb84a270f9123074caf39eb72aaf6f478ab597c6f81435e4b.js?body=1:251)
processResponse @ rails-ujs.self-551fbd47b981dacbb84a270f9123074caf39eb72aaf6f478ab597c6f81435e4b.js?body=1:268
(anonymous) @ rails-ujs.self-551fbd47b981dacbb84a270f9123074caf39eb72aaf6f478ab597c6f81435e4b.js?body=1:196
xhr.onreadystatechange @ rails-ujs.self-551fbd47b981dacbb84a270f9123074caf39eb72aaf6f478ab597c6f81435e4b.js?body=1:251
XMLHttpRequest.send (async)
Rails.ajax @ rails-ujs.self-551fbd47b981dacbb84a270f9123074caf39eb72aaf6f478ab597c6f81435e4b.js?body=1:212
Rails.handleRemote @ rails-ujs.self-551fbd47b981dacbb84a270f9123074caf39eb72aaf6f478ab597c6f81435e4b.js?body=1:568
(anonymous) @ rails-ujs.self-551fbd47b981dacbb84a270f9123074caf39eb72aaf6f478ab597c6f81435e4b.js?body=1:169
```
### System configuration
**Rails version**:
```ruby
gem 'rails',                    '~> 5.2.0'
gem 'jquery-rails',               '~> 4.3', '>= 4.3.3'
```
**Ruby version**:
```ruby
ruby '2.5.0'
```

**Comment by y-yagi** [comment]
An error in JavaScript means that there is some problem with your application. 
Please check the application that there are any places where syntax error occurs. 

Also, please use the [mailing list](https://groups.google.com/forum/#!forum/rubyonrails-talk) or StackOverflow for questions/help, where a wider community will be able to help you. We reserve the issues tracker for issues only.

**Comment by koenhandekyn** [comment]
actually i think i just found the issue. the server side response get's rendered with the HTML layout for some funky reason ?? adding layout false 'fixes' the issue ... but as far as i know it should not be the case that a js response get's wrapped inside the layout ???

`format.js { render 'update', layout: false }`

**Comment by rickychilcott** [comment]
Thanks for sharing this @koenhandekyn; I spent last Friday and part of yesterday trying to get this figured out. I noticed the entire layout seemed to be getting rendered which I thought was weird. This is definitely a "gotcha" and should be reopened.

**Comment by koenhandekyn** [comment]
i actually have more details with respect to my issue. i'm using SLIM as a template language and my application layout was called application.slim instead of application.html.slim. because it was called application.slim it gets applied to all formats. renaming it to application.html.slim fixed the issue more fundamentally.

**Comment by rickychilcott** [comment]
Gotcha. Same issue for me, I suppose. I'm using haml and my files are `.haml` and not `.html.haml` 😏 

Thanks!

**Comment by franks921** [comment]
I've also been running into this issue and I was wrecking my brain when I remembered `data-type`. I vaguely recalled that `rails-ujs` prefers `script` and I checked the [source](https://github.com/rails/rails/blob/master/actionview/app/assets/javascripts/rails-ujs/features/remote.coffee#L25).

In short, adding `'data-type': 'html'` seemed to solve my particular problem.

Here are some additional notes to help anyone else experiencing this issue.

* If you want to avoid adding `data-type` make sure you escape your HTML in the partial using `j` / `escape_javascript` if you are using jQuery.
* In my many years using Rails I found it's just easier using `js.erb` instead of `js.slim` / `js.haml`

**Comment by Chadh13** [comment]
After wrestling with this for much of the weekend, I believe that I understand the cause of this error.

### tl;dr:
You have a Content-Type mismatch between what you are returning and what RailsUJs is expecting to get back.

### Explanation:
I am on Rails 5.2, using a form generated by using the `form_with` view helper, and submitting via AJAX. Additionally, this form has Rails-UJs event handlers attached to it which fire StimulusJS controller events when the callbacks return.

Here is my request cycle:
1) Submit form with AJAX.
2) Render the html from a partial within the controller, that gets passed to the response body by default.
3) The Rails-UJs event listener callback returns, and it fires the StimulusJS Controller Action.
4) The StimulusJS Controller Action handles inserting the HTML within the response body into the DOM.

Note: I am not utilizing `.js.erb/.slim/.haml` files.

The error is occurring within the ternary on this line:
https://github.com/rails/rails/blob/014aa10fc0c2e1431235134c5f94c467e8c17cbd/actionview/app/assets/javascripts/rails-ujs/utils/ajax.coffee#L18

The problem is that `xhr.getResponseHeader('Content-Type')` is inferred first by the Content-Type which you are responding to within the `respond_to`, and then by the Content-Type of the content you are generating within the `render` method:
https://github.com/rails/rails/blob/0196551e6039ca864d1eee1e01819fcae12c1dc9/actionpack/lib/action_controller/metal/mime_responds.rb#L212
https://github.com/rails/rails/blob/f45e23ead96507c2bb582c223f2615ca71244312/actionpack/lib/abstract_controller/rendering.rb#L29

Furthermore, the `Accept-Headers` explicitly ask for some sort of JS `Content-Type`, and exclude `text/html` as the option specifically for the `data-type: html` option. This is why @franks921's suggestion above got rid of the error.
https://github.com/rails/rails/blob/014aa10fc0c2e1431235134c5f94c467e8c17cbd/actionview/app/assets/javascripts/rails-ujs/utils/ajax.coffee#L13

Utilizing `.js.erb/.slim/.haml` files does in fact return an expected JS-related `Content-Type` because they respond with the unexecuted JavaScript that is the `.js.erb` file, and then execute the JS in the browser after the request has been returned.

Meanwhile, if you render anything except for raw unexecuted JavaScript in your response, such as plain HTML; then you are returning an HTML response with a Content-Type of JS.

In these cases, if you explicitly include `content_type: 'text/html'` in your render call then the RailsUJs error goes away:
- `render  partial: 'flashes'` executes correctly but produces the RailsUJs error in the console.
- `render  partial: 'flashes', content_type: 'text/html'` executes correctly without the console error.

Even after diving into this, I'm not sure I have a strong plan, idea, or opinion about if or how this should be solved within the RailsUJs code itself. I don't fully comprehend the importance and reasoning for splitting out the `script` and the `html` AcceptHeader options within RailsUJs.

It appears to be related to Content Security Policy and potentially CSRF/XSS (html sanitization):
https://github.com/rails/rails/blob/014aa10fc0c2e1431235134c5f94c467e8c17cbd/actionview/app/assets/javascripts/rails-ujs/utils/ajax.coffee#L68

In any case, perhaps this new information may give @y-yagi or another team member some additional context if there is a problem to solve within the lib itself.

**Comment by Unkas82** [comment]
> actually i think i just found the issue. the server side response get's rendered with the HTML layout for some funky reason ?? adding layout false 'fixes' the issue ... but as far as i know it should not be the case that a js response get's wrapped inside the layout ???
> 
> `format.js { render 'update', layout: false }`

Amazing! U've saved my day!

**Comment by thomasbromehead** [comment]
@koenhandekyn You saved me a lot of searching...!!!
